client/client-lib/src/mint/backup.rs

```diff
@@ -0,0 +1,218 @@
+    }
+
+    /// Static version of [`Self::get_derived_backup_encryption_key`] for testing without creating whole `MintClient`
+    fn get_derived_backup_encryption_key_static(secret: &DerivableSecret) -> aead::LessSafeKey {
+        // TODO: Do we need that one derivation level? This key is already derived for the mint itself, and internally another kdf will be done with key type tag.
+        aead::LessSafeKey::new(
+            secret
+                .child_key(ChildId(BACKUP_CHILD_ID))
+                .to_chacha20_poly1305_key(),
+        )
+    }
+
+    /// Static version of [`Self::get_derived_backup_secret_key`] for testing without creating whole `MintClient`
+    fn get_derived_backup_signing_key_static(secret: &DerivableSecret) -> secp256k1_zkp::KeyPair {
+        // TODO: Do we need that one derivation level? This key is already derived for the mint itself, and internally another kdf will be done with key type tag.
+        secret
+            .child_key(ChildId(BACKUP_CHILD_ID))
+            .to_secp_key(&Secp256k1::<secp256k1::SignOnly>::gen_new())
+    }
+
+    fn get_derived_backup_encryption_key(&self) -> aead::LessSafeKey {
+        Self::get_derived_backup_encryption_key_static(&self.secret)
+    }
+
+    fn get_derived_backup_signing_key(&self) -> secp256k1::KeyPair {
+        Self::get_derived_backup_signing_key_static(&self.secret)
+    }
+
+    async fn prepare_plaintext_ecash_backup(&self) -> PlaintextEcashBackup {
+        let mut dbtx = self.start_dbtx();
+
+        let notes = self.get_available_notes(&mut dbtx);
+        let last_idx = self.get_last_note_index(&mut dbtx);
+
+        PlaintextEcashBackup { notes, last_idx }
+    }
+
+    async fn prepare_ecash_backup(&self) -> Result<EcashBackup> {
+        let plaintext = self.prepare_plaintext_ecash_backup().await;
+        plaintext.encrypt_to(&self.get_derived_backup_encryption_key())
+    }
+
+    async fn upload_ecash_backup(&self, backup: EcashBackup) -> Result<()> {
+        let backup_request = backup.into_backup_request(&self.get_derived_backup_signing_key())?;
+        self.context
+            .api
+            .upload_ecash_backup(&backup_request)
+            .await?;
+        Ok(())
+    }
+}
+
+/// Snapshot of a ecash state (notes)
+///
+/// Used to speed up and improve privacy of ecash recovery,
+/// by avoiding scanning the whole history.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct PlaintextEcashBackup {
+    notes: TieredMulti<SpendableNote>,
+    last_idx: NoteIndex,
+}
+
+impl PlaintextEcashBackup {
+    /// Align an ecoded message size up for better privacy
+    fn get_alignment_size(len: usize) -> usize {
+        // TODO: should we align to power of 2 instead?
+        let padding_alignment = 16 * 1024;
+        ((len.saturating_sub(1) / padding_alignment) + 1) * padding_alignment
+    }
+
+    fn bincode_options() -> impl bincode::Options {
+        bincode::DefaultOptions::new()
```

Comment:
I think we want to get rid of bincode eventually since we need the `Encodable` API anyway. Why not just go with `Encodable`/`Decodale` here?

Reply:
done